### PR TITLE
Initialize AsyncLLMEngine bg loop correctly

### PR DESCRIPTION
**Pull request by Yard1**
Fixes wrong asyncio event loop being used when ran through uvicorn/FastAPI.

Fixes https://github.com/vllm-project/vllm/pull/880#issuecomment-1705015828

@zhuohan123

### CODE REVIEW

File: vllm/engine/async_llm_engine.py [hunk 1]
```diff
@@ -155,9 +155,13 @@ def __init__(self,
...
+            self.start_background_loop()
 
-    def _start_background_loop(self) -> None:
+    @property
+    def is_running(self) -> bool:
+        return self.background_loop is not None
+
+    def start_background_loop(self) -> None:
         """Start the background loop."""
         if self.background_loop is not None:
```
Comment (Yard1) [inline vllm/engine/async_llm_engine.py:166]:
```suggestion
        if self.is_running:
```

**Review by zhuohan123 — APPROVED** [review]
LGTM! Thanks for the quick fix!

**Comment by esmeetu** [comment]
@Yard1 @zhuohan123 openai server cannot accept new request when abort request or request unexpectedly disconnected. pls check again.

**Comment by Yard1** [comment]
@esmeetu I just checked the current master and it seems to be working fine - can you check again and tell me exactly what steps you took and requests you sent?

**Comment by esmeetu** [comment]
@Yard1 
Related code:
def _raise_exception_on_finish(task: asyncio.Task) -> None:
    try:
        task.result()
    except Exception as e:
        raise RuntimeError("Task finished unexpectedly.") from e
    raise RuntimeError("Task finished unexpectedly.")

When i abort request, it will raise RuntimeError("Task finished unexpectedly."). And background loop seems dead, cannot run new request task. I will reproduce later again.

**Comment by Yard1** [comment]
@esmeetu Can you try with https://github.com/vllm-project/vllm/pull/957 ?

**Comment by esmeetu** [comment]
@Yard1 Same as before using latest main branch.

Correct the reproduce step:

SEND 100 requests simultaneously, then cancel all. Single request works right.

Error message:

Exception in callback _raise_exception_on_finish(<Task finishe...239e8d61324')>) at /home/dell/vllm/vllm/engine/async_llm_engine.py:49
handle: <Handle _raise_exception_on_finish(<Task finishe...239e8d61324')>) at /home/dell/vllm/vllm/engine/async_llm_engine.py:49>
Traceback (most recent call last):
  File "/home/dell/vllm/vllm/engine/async_llm_engine.py", line 51, in _raise_exception_on_finish
    task.result()
  File "/home/dell/vllm/vllm/engine/async_llm_engine.py", line 214, in run_engine_loop
    await self.engine_step()
  File "/home/dell/vllm/vllm/engine/async_llm_engine.py", line 192, in engine_step
    self.request_streams[request_id].put(request_output)
KeyError: 'cmpl-f1a1a40a81314a42b1305239e8d61324'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/dell/miniconda3/envs/ray/lib/python3.10/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/home/dell/vllm/vllm/engine/async_llm_engine.py", line 53, in _raise_exception_on_finish
    raise RuntimeError("Task finished unexpectedly.") from e
RuntimeError: Task finished unexpectedly.

**Comment by Yard1** [comment]
Thank you, I will try to reproduce tomorrow

**Comment by Yard1** [comment]
@esmeetu can you share what request (including sampling params) you were sending?

**Comment by esmeetu** [comment]
@Yard1 all requests like this(Humaneval benchmark),
Received request cmpl-4df9e168eeab4fa3b11f49cc7fe48b70: prompt: 'Below is an instruction that describes a task. Write a response that appropriately completes the request.\n\n\n### Instruction:\nCreate a Python script for this problem:\nfrom typing import List\n\n\ndef has_close_elements(numbers: List[float], threshold: float) -> bool:\n    """ Check if in given list of numbers, are any two numbers closer to each other than\n    given threshold.\n    >>> has_close_elements([1.0, 2.0, 3.0], 0.5)\n    False\n    >>> has_close_elements([1.0, 2.8, 3.0, 4.0, 5.0, 2.0], 0.3)\n    True\n    """\n\n\n### Response:', sampling params: SamplingParams(n=1, best_of=1, presence_penalty=0.0, frequency_penalty=0.0, temperature=0.0, top_p=1.0, top_k=-1, use_beam_search=False, length_penalty=1.0, early_stopping=False, stop=[], ignore_eos=False, max_tokens=2048, logprobs=None)


And i use multiprocessing lib to request api endpoint simultaneously.

**Pull request merged**